proposals/2025-02-24_secret_providers.md

```diff
@@ -0,0 +1,147 @@
+* Default values
+* De-duplication and re-using common config values for secret provider configs
+  * This will be left to a follow up proposal if needed
+
+## How
+
+
+### Configuration
+
+Wherever a `<secret>` type is present in the configuration files, we will allow users to specify a map specifying how to fetch from a secret provider. The generic format would be
+
+```
+secret_field:
+  provider:  <type of the provider>
+  <property1>: <value1>
+  ...
+  <propertyN>: <valueN>
+```
+
+For example when specifying a password fetched from the kubernetes provider with an id of `pass2` in namespace `ns1` for the HTTP passsword field it would look like this:
+
+```
+        password:
+          provider: kubernetes
+          namespace: ns1
+          secret_id: pass2
+```
+
+### Inline secrets
+
+When specifying secrets inline in the config file, a string can be passed in as usual for backwards compatibility.
+```
+        password: my_important_secret
+```
+
+### Error handling
+
+We can classify all errors stemming from secret provider failures into 2 cases.
+
+The first case is that we have not gotten any secret value since startup. In this case we should not initialize any component that mentions this faulty secret, log this failure and schedule a retry to get this secret value and initialize the component.
+
+The second case is that we already have a secret value, but refreshing it has resulted in an error. In this case we should keep the component that uses this secret running with the potentially stale secret, schedule a retry and 
+
+### Metrics
+
+Metrics should be present to help users identify the errors mentioned above (in addition to logs). Therefore the following metrics should be reported per secret present in the config file:
+
+The time since the last successful secret fetch
+
+```
+prometheus_remote_secret_last_successful_fetch_seconds{provider="kubernetes", secret_id="pass1"} 15
+```
+
+A state enum describing in which error condition the secret is in:
+* error: there has been no successful request and no secret has been retrieved
+* stale: a request has succeeded but the latest request failed
+* none: the last request was succesful
+
+```
+# HELP prometheus_remote_secret_state Describes the current state of a remotely fetched secret.
+# TYPE prometheus_remote_secret_state gauge
+prometheus_remote_secret_state{provider="kubernetes", secret_id="pass1", state="none"} 0
+prometheus_remote_secret_state{provider="kubernetes", secret_id="auth_token", state="stale"} 1
+prometheus_remote_secret_state{id="myk8secrets", secret_id="pass2", state="error"} 2
+```
+
+### Nested secrets
+
+Secret providers might require secrets to be configured themselves. We will allow secrets to be passed in to secret providers.
+
+```
+...
+        password: !secret id=bootstrapped secret_id=pass1
+          provider: bootstrapped
+          secret_id: pass1
+          auth_token:
+            provider: kubernetes
+            secret_id: auth_token
```

Comment:
```suggestion
...
        password:
          provider: bootstrapped
          secret_id: pass1
          auth_token:
            provider: kubernetes
            secret_id: auth_token
```
I think you forgot to remove the tag here.

Reply:
Fixed as well thank you